docs/tutorials/training_agents/deep_rl_tutorial.py

```diff
@@ -0,0 +1,1069 @@
+# main network's.
+# Lastly, since an agent needs to do exploration using epsilon-greedy, there is a way for updating the exploration rate,
+# i.e. ``updating epsilon``.
+#
+class ReplayMemory:
+    """Implements a circular replay memory object based on a deque."""
+
+    def __init__(self, params):
+        """
+        Initialize the replay memory.
+        Args:
+            params: Hyperparameters
+        """
+        self._params = params
+        self._buffer = collections.deque([], maxlen=params.capacity)
+
+    def push(
+        self,
+        obs: gymnasium.wrappers.LazyFrames,
+        action: int,
+        reward: gymnasium.core.SupportsFloat,
+        next_obs: gymnasium.wrappers.LazyFrames,
+        done: bool,
+    ):
+        """
+        Add a transition to the replay memory. When the buffer is full,
+        the oldest transitions are replaced with new ones.
+
+        Args:
+            obs: Agent's observation
+            action: Executed action.
+            reward: Reward received.
+            next_obs: Resulting observation.
+            done: Terminal state.
+        """
+        self._buffer.append((obs, action, reward, next_obs, int(done)))
+
+    def sample(self) -> tuple:
+        """
+        Sample a minibatch of transitions.
+
+        Raises:
+             ValueError: if not enough transitions exist to sample.
+
+         Returns:
+             5-tuple of obs, actions, rewards, next_obs, dones
+
+        """
+        if len(self._buffer) < self._params.batch_size:
+            raise ValueError("Not enough transitions to sample a minibatch")
+
+        sample = random.sample(self._buffer, self._params.batch_size)
+        return tuple(zip(*sample))
+
+
+class Network(nn.Module):
+    """
+    Class implementation of the Deep-Q-Network architecture, where
+    it outputs return distributions instead of action-values.
+    """
+
+    def __init__(self, params):
+        """Initialize the network. Expects hyperparameters object."""
+        super().__init__()
+        self._params = params
+
+        if self._params.image_obs:
+            self._convolutional = nn.Sequential(
+                nn.Conv2d(4, 32, (8, 8), (4, 4), 0),
+                nn.ReLU(),
+                nn.Conv2d(32, 64, (4, 4), (2, 2), 0),
+                nn.ReLU(),
+                nn.Conv2d(64, 64, (3, 3), (1, 1), 0),
+                nn.ReLU(),
+                nn.Flatten(),
+            )
+        else:
+            self._convolutional = nn.Sequential()
+
+        in_features = self._output_size()
+        out_features = self._params.n_actions * self._params.n_atoms
+        n_hidden_units = self._params.n_hidden_units
+
+        self._head = torch.nn.Sequential(
+            nn.Linear(in_features=in_features, out_features=n_hidden_units),
+            nn.ReLU(),
+            nn.Linear(in_features=n_hidden_units, out_features=out_features),
+        )
+
+    def _output_size(self) -> int:
```

Comment:
I am not sure how it could be computed from only `env.action_space.shape` or `env.observation_space.shape` without going through the steps made by the `Conv2d` layer, that is `kerner_size`, `stride` `padding` etc.

Reply:
doesn't `output_size == action_space.shape[0]`??

Reply:
If we are talking about the output of the network, then in C51 it is `action_space.shape[0] * n_atoms`.

However, note the two lines 
```
in_features = self._output_size()
out_features = self._params.n_actions * self._params.n_atoms
```
where the method `_output_size` is used to compute the output size of the flatten layer. Maybe the choice of method name is not good?

Reply:
"`in_features = self._output_size()`" I very confused by what I am reading here, maybe you need to rename the function `_output_size` to something else, or add some more comments

Reply:
Suspected that. Changed the method name to:
```    
def _fc_input_size(self) -> int:
        """Compute size of input to first linear layer."""
```

Reply:
isn't the case that:
`in_features = env.observation_space.shape[0]`

Reply:
Partially true.

If the vision part of the network is not utilized, then indeed `in_features = env.observation_space.shape[0]`, and this is exactly what is returned by `_fc_input_size`, since the convolutional sequential is empty, so the observation is returned as is.

But if the vision part is active, then it depends on the observation shape and convolutional operations.
Assuming that `env.observation_space.shape == (4, 84, 84)` and the current convolutional structure, then the size is computed manually as follows:
the output shape of the first conv2d layer is `(32, 20, 20)`,
the output shape of the second conv2d layer is `(64, 9, 9)` and 
the output shape of the third conv2d layer is `(64, 7, 7)`
which will produce `64 * 7 * 7 = 3136` weights at the flatten layer.

But we can't always assume that the `env.observation_space.shape == (4, 84, 84)` or that the network has this structure in general. So I prefer computing it using the current method.